File: src/canvas.cpp
#include "canvas.h"
#include "comparinator.h"
#include "color.h"
#include "pch.h"

Canvas::Canvas(int width, int height) {
	this->mbrWidth = width;
	this->mbrHeight = height;
	this->mbrGrid.clear();
	for (int i = 0; i < mbrWidth; i++)
	{
		std::vector<Color> column;
		mbrGrid.push_back(column);
		for (int j = 0; j < mbrHeight; j++)
		{
			mbrGrid[i].push_back(Color(0, 0, 0));
		}
	}
};

bool Canvas::isClean() {
	Comparinator ce = Comparinator();
	Color black = Color(0, 0, 0);
	bool clean = true;
	for (int i = 0; i < mbrGrid.size(); i++)
	{
		for (int j = 0; j < mbrGrid[0].size(); j++)
		{
			if (!ce.checkTuple(mbrGrid[i][j], black))
			{
				clean = false;
				break;
			}
		}
	}
	return clean;
};

bool Canvas::inBounds(int x,int y) {
	return mbrGrid.size() > 0 && mbrGrid[0].size() > 0 && x >= 0 && x < mbrGrid.size() && y >= 0 && y < mbrGrid[0].size();
}

void Canvas::setPixel(int x, int y, Color c) {
	if (inBounds(x,y)) { mbrGrid[x][y] = c; }
}

void Canvas::fill(Color x) {
	for (int i = 0; i < mbrGrid.size(); i++)
	{
		for (int j = 0; j < mbrGrid[i].size(); j++)
		{
			setPixel(i, j, x);
		}
	}
}

Color Canvas::getPixel(int x, int y) {
	if (inBounds(x, y)) { return mbrGrid[x][y]; }
	return Color(0,0,0);
}

std::string Canvas::getPPM()
{
	//std::string ppm = "P3\n" + std::to_string(w) + " " + std::to_string(h) + "\n" + "255\n";
	std::string* ppm = new std::string("P3\n" + std::to_string(mbrWidth) + " " + std::to_string(mbrHeight) + "\n" + "255\n");
	std::string buffer = "";
	int cnr = 0;
	int cng = 0;
	int cnb = 0;
	// width = row & height = column
	for (int j = 0; j < mbrHeight; j++)
	{
		for (int i = 0; i < mbrWidth; i++)
		{
			cnr = std::min((int)std::round((255 * mbrGrid[i][j].mbrRed)), 255);
			cnr = cnr > 0 ? cnr : 0;
			std::string clampedNormalizedRed = std::to_string(cnr);
			if (buffer.size() + clampedNormalizedRed.size() > getPPMLineWidth())
			{
				*ppm += buffer;
				*ppm = isspace((*ppm)[(*ppm).size() - 1]) ? (*ppm).substr(0, (*ppm).size() - 1) : *ppm;
				buffer = "\n" + clampedNormalizedRed + " ";
			}
			else
			{
				buffer += clampedNormalizedRed + " ";
			}
			cng = std::min((int)std::round((255 * mbrGrid[i][j].mbrGreen)), 255);
			cng = cng > 0 ? cng : 0;
			std::string clampedNormalizedGreen = std::to_string(cng);
			if (buffer.size() + clampedNormalizedGreen.size() > getPPMLineWidth())
			{
				*ppm += buffer;
				*ppm = isspace((*ppm)[(*ppm).size() - 1]) ? (*ppm).substr(0, (*ppm).size() - 1) : *ppm;
				buffer = "\n" + clampedNormalizedGreen + " ";
			}
			else
			{
				buffer += clampedNormalizedGreen + " ";
			}
			cnb = std::min((int)std::round((255 * mbrGrid[i][j].mbrBlue)), 255);
			cnb = cnb > 0 ? cnb : 0;
			std::string clampedNormalizedBlue = std::to_string(cnb);
			if (buffer.size() + clampedNormalizedBlue.size() > getPPMLineWidth())
			{
				*ppm += buffer;
				*ppm = isspace((*ppm)[(*ppm).size() - 1]) ? (*ppm).substr(0, (*ppm).size() - 1) : *ppm;
				buffer = "\n" + clampedNormalizedBlue + " ";
			}
			else
			{
				buffer += clampedNormalizedBlue + " ";
			}
		}
		*ppm += buffer;
		*ppm = isspace((*ppm)[(*ppm).size() - 1]) ? (*ppm).substr(0, (*ppm).size() - 1) : *ppm;
		buffer = "\n";
		//std::cout << "rendered line:" << std::to_string(j) << std::endl;
	}
	// cutoff extra whitespace
	*ppm = isspace((*ppm)[(*ppm).size() - 1]) ? (*ppm).substr(0, (*ppm).size() - 1) : *ppm;
	*ppm += "\n";
	return *ppm;
}

void Canvas::save() {
	std::ofstream file(getPPMFilename(true));
	std::cout << "to saving ..." << std::endl;
	std::string data = getPPM();
	if (file.is_open()) {
		file << data;
	}
	file.close();
	std::cout << "saved" << std::endl;
}

File: src/Canvas.cs
using LibColor;
using LibComparinator;
using LibProjectMeta;
namespace LibCanvas;
public class Canvas {
	public int mbrWidth;
	public int mbrHeight;
	public List<List<Color>> mbrGrid;
	public Canvas(int width, int height) {
		mbrWidth = width;
		mbrHeight = height;
		mbrGrid.Clear();
		for (int i = 0; i < mbrWidth; ++i)
		{
			mbrGrid.Add(new List<Color>());
			for (int j = 0; j < mbrHeight; ++j)
			{
				mbrGrid[i].Add(new Color(0, 0, 0));
			}
		}
	}
	public Color getPixel(int x, int y) {
		if (inBounds(x, y)) { return mbrGrid[x][y]; }
		return new Color(0,0,0);
	}
	public String getPPM()
	{
		ProjectMeta varPM = new ProjectMeta();
		//String ppm = "P3\n" + std::to_string(w) + " " + std::to_string(h) + "\n" + "255\n";
		String ppm = "P3\n" + mbrWidth.ToString() + " " + mbrHeight.ToString() + "\n" + "255\n";
		String buffer = "";
		int cnr = 0;
		int cng = 0;
		int cnb = 0;
		// width = row & height = column
		for (int j = 0; j < mbrHeight; ++j)
		{
			for (int i = 0; i < mbrWidth; ++i)
			{
				cnr = Math.Min((int)Math.Round((255 * mbrGrid[i][j]._fieldRed)), 255);
				cnr = cnr > 0 ? cnr : 0;
				String clampedNormalizedRed = cnr.ToString();
				if (buffer.Count() + clampedNormalizedRed.Count() > varPM.getPPMLineWidth())
				{
					ppm += buffer;
					ppm = char.IsWhiteSpace(ppm[ppm.Count() - 1]) ? ppm.Substring(0, ppm.Count() - 1) : ppm;
					buffer = "\n" + clampedNormalizedRed + " ";
				}
				else
				{
					buffer += clampedNormalizedRed + " ";
				}
				cng = Math.Min((int)Math.Round((255 * mbrGrid[i][j]._fieldGreen)), 255);
				cng = cng > 0 ? cng : 0;
				String clampedNormalizedGreen = cng.ToString();
				if (buffer.Count() + clampedNormalizedGreen.Count() > varPM.getPPMLineWidth())
				{
					ppm += buffer;
					ppm = char.IsWhiteSpace(ppm[ppm.Count() - 1]) ? ppm.Substring(0, ppm.Count() - 1) : ppm;
					buffer = "\n" + clampedNormalizedGreen + " ";
				}
				else
				{
					buffer += clampedNormalizedGreen + " ";
				}
				cnb = Math.Min((int)Math.Round((255 * mbrGrid[i][j]._fieldBlue)), 255);
				cnb = cnb > 0 ? cnb : 0;
				String clampedNormalizedBlue = cnb.ToString();
				if (buffer.Count() + clampedNormalizedBlue.Count() > varPM.getPPMLineWidth())
				{
					ppm += buffer;
					ppm = char.IsWhiteSpace(ppm[ppm.Count() - 1]) ? ppm.Substring(0, ppm.Count() - 1) : ppm;
					buffer = "\n" + clampedNormalizedBlue + " ";
				}
				else
				{
					buffer += clampedNormalizedBlue + " ";
				}
			}
			ppm += buffer;
			ppm = char.IsWhiteSpace(ppm[ppm.Count() - 1]) ? ppm.Substring(0, ppm.Count() - 1) : ppm;
			buffer = "\n";
			//Console.WriteLine("rendered line:" << std::to_string(j));
		}
		// cutoff extra whitespace
		ppm = char.IsWhiteSpace(ppm[ppm.Count() - 1]) ? ppm.Substring(0, ppm.Count() - 1) : ppm;
		ppm += "\n";
		return ppm;
	}
	public bool isClean() {
		Comparinator ce = new Comparinator();
		Color black = new Color(0, 0, 0);
		bool clean = true;
		for (int i = 0; i < mbrGrid.Count(); ++i)
		{
			for (int j = 0; j < mbrGrid[0].Count(); ++j)
			{
				if (!ce.CheckTuple(mbrGrid[i][j], black))
				{
					clean = false;
					break;
				}
			}
		}
		return clean;
	}
	public bool inBounds(int x, int y) {
		return mbrGrid.Count() > 0 && mbrGrid[0].Count() > 0 && x >= 0 && x < mbrGrid.Count() && y >= 0 && y < mbrGrid[0].Count();
	}
	public void setPixel(int x, int y, Color c) {
		if (inBounds(x,y)) { mbrGrid[x][y] = c; }
	}
	public void fill(Color x)  {
		for (int i = 0; i < mbrGrid.Count(); ++i)
		{
			for (int j = 0; j < mbrGrid[i].Count(); ++j)
			{
				setPixel(i, j, x);
			}
		}
	}
	public void save(){
		ProjectMeta varPm = new ProjectMeta();
		// std::ofstream file(varPm.getPPMFilename(true));
		// Console.WriteLine("to saving ...");
		// String data = getPPM();
		// if (file.is_open()) {
		// 	file << data;
		// }
		// file.close();
		using (StreamWriter writer = new StreamWriter(varPm.getPPMFilename(true))) {
			writer.WriteLine(getPPM());
		}
		Console.WriteLine("saved");
	}
};